docs/design-proposals/2023-05 vParquet3.md

```diff
@@ -0,0 +1,179 @@
+* [Implementation](#implementation)
+  * [Schema with spare columns](#schema-with-spare-columns)
+  * [Configuration and overrides](#configuration-and-overrides)
+  * [Compaction](#compaction)
+  * [Tooling](#tooling)
+
+## Context
+The current Parquet schema stores span attributes in the repeated group `Attrs`.
+The group contains a single column for all attribute keys and separate columns for the corresponding values (depending on the value type).
+The size of the span attribute columns combined contribute to 30-70% of the overall block size.
+To search for a specific attribute, it is necessary to access this large chunk of data.
+The analysis of Tempo blocks from different use cases has further revealed that the 10 most frequently used span attributes are responsible for over 50% of the overall attribute size.
+
+A benchmark with a schema that stores the 5 most frequently occurring attributes in dedicated columns showed significant performance improvements:
+
+* The block size decreased by ~10%
+* Compressed size of the generic attribute columns was reduced by ~40%
+* The search speed for queries without matches improved by ~80% for attributes with dedicated columns and ~70% for generic attributes columns
+* The search speed for queries returning matches improved by ~50% for attributes with dedicated columns and ~30% for generic attribute columns
+
+The amount and kind of span attributes varies a lot between use cases and the above benchmark was conducted with tracing data that captured a lot of span attributes.
+Therefore, the search speed improvements that can be achieved by storing span attributes in dedicated columns,
+can deviate from the above results.
+However, the data still shows that dedicated columns for span attributes that can be tuned to a specific use case can improve Tempo’s search performance significantly.
+This document describes how dedicated columns for span attributes will be implemented in Tempo.
+
+## Goals and requirements
+
+There are different ways dedicated columns for attributes can be added to Tempo's block format and how they are configured and tuned for the specific workload.
+The main goal for the next iteration of the block format `vParquet3` is to be able to manually configure dedicated attribute columns for very heavy workloads.
+To keep the scope small only the following features are implemented for now:
+
+* Attributes with string values can be promoted to dedicated columns on resource and span level
+* Tenants can manually configure which attributes are promoted to dedicated columns
+* Support for changing dedicated column configurations
+  * Compactor treats different configs as different encodings
+  * Queriers read blocks with different configured columns
+* Provide tooling and documentation that enables tenants to find good attributes to promote to dedicated columns
+  * Find attributes that contribute the most to block size
+  * (Stretch) Find attributes that are search the most by
```

Comment:
by .... ?

This could be nice. Maybe some kind of query-frontend log analysis tool? I'd even be satisfied for some documented Loki queries that show what are the most commonly searched for attributes.

Overall tho I agree this is stretch and shouldn't hold up the project.

Reply:
Ah, apologies. I thought this was an incomplete sentence and you meant to type something after "by". I understand now.